File: api/app/parsers/package_repository_resolvers/npm.py
import httpx
from typing import Optional
from app.parsers.package_repository_resolvers.base import register_package_repository_resolver


@register_package_repository_resolver("npm")
async def get_npm_repo_url(name: str) -> Optional[str]:
    async with httpx.AsyncClient() as client:
        response = await client.get(f"https://registry.npmjs.org/{name}")

        if response.status_code == 200:
            repository = response.json().get("repository", {})
            url = repository.get("url") or ""

            return url.replace("git+", "").replace(".git", "").replace("git://", "https://")

    return None

File: background-worker/internal/dependency/resolvers/npm.go
package resolvers

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"strings"
)

func GetNpmRepoURL(client *http.Client, baseURL string) func(ctx context.Context, name string) (string, error) {
	return func(ctx context.Context, name string) (string, error) {
		url := fmt.Sprintf("%s/%s", baseURL, name)
		req, _ := http.NewRequestWithContext(ctx, "GET", url, nil)

		resp, err := client.Do(req)
		if err != nil {
			return "", err
		}

		defer func() {
			if err := resp.Body.Close(); err != nil {
				log.Printf("failed to close response body: %v", err)
			}
		}()

		if resp.StatusCode != 200 {
			return "", nil
		}

		var data struct {
			Repository struct {
				URL string `json:"url"`
			} `json:"repository"`
		}

		if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
			return "", err
		}

		repoURL := data.Repository.URL
		repoURL = strings.ReplaceAll(repoURL, "git+", "")
		repoURL = strings.ReplaceAll(repoURL, ".git", "")
		repoURL = strings.ReplaceAll(repoURL, "git://", "https://")

		return repoURL, nil
	}
}
